src/warc2zim/main.py

```diff
@@ -219,64 +259,204 @@ def add_remote_or_local(self, filename):
+
+        if not self.favicon_url:
+            # find icon
+            icon = soup.find("link", rel="shortcut icon")
+            if not icon:
+                icon = soup.find("link", rel="icon")
+
+            if icon:
+                self.favicon_url = urljoin(self.main_url, icon.attrs["href"])
+            else:
+                self.favicon_url = urljoin(self.main_url, "/favicon.ico")
+
+        if not self.language:
+            # HTML5 Standard
+            lang_elem = soup.find("html", attrs={"lang": True})
+            if lang_elem:
+                self.language = lang_elem.attrs["lang"]
+                return
+
+            # W3C recommendation
+            lang_elem = soup.find(
+                "meta", {"http-equiv": "content-language", "content": True}
+            )
+            if lang_elem:
+                self.language = lang_elem.attrs["content"]
+                return
+
+            # SEO Recommendations
+            lang_elem = soup.find("meta", {"name": "language", "content": True})
+            if lang_elem:
+                self.language = lang_elem.attrs["content"]
+                return
+
+    def generate_all_articles(self):
+        if self.favicon_url:
+            yield FaviconArticle(self.favicon_url)
+
+        # add replay system
+        for article in self.replay_articles:
+            yield article
+
+        for record in self.iter_warc_records():
+            yield from self.articles_for_warc_record(record)
+
+        # process revisits, headers only
+        for url, record in self.revisits.items():
+            if url not in self.indexed_urls:
+                logger.debug(
+                    "Adding revisit {0} -> {1}".format(
+                        url, record.rec_headers["WARC-Refers-To-Target-URI"]
+                    )
+                )
+                yield WARCHeadersArticle(record)
+                self.indexed_urls.add(url)
+
+    def articles_for_warc_record(self, record):
+        url = record.rec_headers["WARC-Target-URI"]
+        if url in self.indexed_urls:
+            logger.warning("Skipping duplicate {0}, already added to ZIM".format(url))
+            return
+
+        # if not include_all, only include urls from main_url domain or subdomain
+        if not self.include_all:
+            parts = urlsplit(url)
+            if not any(
+                parts.netloc.endswith(domain) for domain in self.include_domains
+            ):
+                logger.debug("Skipping url {0}, outside included domains".format(url))
+                return
+
+        if record.rec_type != "revisit":
+            yield WARCHeadersArticle(record)
+            payload_article = WARCPayloadArticle(record)
 
-                    if record.rec_type != "revisit":
-                        zimcreator.add_article(WARCHeadersArticle(record))
-                        payload_article = WARCPayloadArticle(record)
+            if len(payload_article.payload) != 0:
```

Comment:
Shouldn't you check the headers before creating the PayloadArticle first? So you can skip early if Content-Length is null. You could still test of the header is not set and that would still save this test for the 99.9999% of the records where we do have a payload.

Reply:
There could be a `Transfer-Encoding: chunked` on the HTTP headers which means we don't know what the payload is until we read it fully.. Though there are ways to compute it by subtracting the current stream position from total length of record, I don't think that's even worth it here. Since we're always reading the payload if it exists, it's not really much of an optimization even then.. I guess would just save a constructor call maybe? Seems a very very minor difference, and there shouldn't be many of these anyway... Could also raise an exception in constructor if payload is 0 but again, seems like its making it more complicated than what is there now : )

Reply:
OK, that's fine then